### ISSUE TITLE
Performance regression with GraalVM CE/EE for Java 16

### ISSUE THREAD

**Issue by plokhotnyuk** [oracle/graal#3349]
Unexpected regression of throughput (down to 3x times) was detected for some simple benchmarks of the [jsoniter-scala](https://github.com/plokhotnyuk/jsoniter-scala) project.

**Steps to reproduce the issue**
1. `git clone --depth 1 https://github.com/plokhotnyuk/jsoniter-scala`
2. `cd jsoniter-scala`
2. `sbt -java-home <path-to-graalvm-ce-home> 'jsoniter-scala-benchmarkJVM/jmh:run benchmark.Int(Reading|Writing).jsoniterScala'`

**Describe GraalVM and your environment:**
 - GraalVM versions: **CE 21.1.0-dev (build 16+36-jvmci-21.1-b03)** and **EE 21.1.0 (build 16.0.1+9-jvmci-21.1-b05)**
 - JDK major version: **16**
 - OS: **Ubuntu 20.04 (Linux 5.4.0-70-generic)**
 - Architecture: **AMD64**

**More details**
Below are results of benchmarks for different Java versions of the same GraalVM CE release:
***Java 16***
```
[info] Benchmark                          Mode  Cnt         Score         Error  Units
[info] IntReading.jsoniterScala          thrpt    5  17477295.379 ± 1172342.120  ops/s
[info] IntWriting.jsoniterScala          thrpt    5  17728870.533 ± 1554436.005  ops/s
[info] IntWriting.jsoniterScalaPrealloc  thrpt    5  20556882.827 ± 3512248.333  ops/s
```
***Java 11***
```
[info] Benchmark                          Mode  Cnt         Score          Error  Units
[info] IntReading.jsoniterScala          thrpt    5  45519033.777 ± 1697668.569  ops/s
[info] IntWriting.jsoniterScala          thrpt    5  43545371.033 ±  3412878.055  ops/s
[info] IntWriting.jsoniterScalaPrealloc  thrpt    5  60051458.540 ± 10401355.468  ops/s
```
***Java 8***
```
[info] Benchmark                          Mode  Cnt         Score         Error  Units
[info] IntReading.jsoniterScala          thrpt    5  43696270.401 ± 3816447.530  ops/s
[info] IntWriting.jsoniterScala          thrpt    5  45415989.394 ± 8868985.986  ops/s
[info] IntWriting.jsoniterScalaPrealloc  thrpt    5  58274592.891 ±  155174.641  ops/s
```

**Comment by plokhotnyuk** [comment]
I got almost the same kind of performance drop with GraalVM EE for Java 16:
![image](https://user-images.githubusercontent.com/890289/115676088-87bd3500-a34f-11eb-8452-c7dcf1ccec39.png)
![image](https://user-images.githubusercontent.com/890289/115676158-99064180-a34f-11eb-8852-19c2addcc863.png)

**Comment by plokhotnyuk** [comment]
Please use following links to see difference of results for all benchmarks between versions of GraalVM:
1) [GraalVM CE Java11 vs. GraalVM CE Java16](https://jmh.morethan.io/?sources=https://plokhotnyuk.github.io/jsoniter-scala/graalvm-ce-java11.json,https://plokhotnyuk.github.io/jsoniter-scala/graalvm-ce-java16.json)
2) [GraalVM EE Java11 vs. GraalVM EE Java16](https://jmh.morethan.io/?sources=https://plokhotnyuk.github.io/jsoniter-scala/graalvm-ee-java11.json,https://plokhotnyuk.github.io/jsoniter-scala/graalvm-ee-java16.json)

**Comment by mur47x111** [comment]
I have looked at `IntReading.jsoniterScala` and it is due to the not-yet-supported intrinsic of `Reference.refersTo` called from `ThreadLocal.get()`. See [JDK-8256999](https://bugs.openjdk.java.net/browse/JDK-8256999), [CheckGraalIntrinsics.java#L407-L410](https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java#L407-L410).

**Comment by mur47x111** [comment]
Should be fixed by 3f23563b101e1dd92ec086136bc86351da17593f . Please reopen if that is not the case.
